http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -2,131 +2,229 @@ package org.broadinstitute.dsde.workbench.leonardo
+}
 import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
 import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, ServiceAccountKey, ServiceAccountKeyId}
 import org.broadinstitute.dsde.workbench.util.Retry
 
-import scala.concurrent.ExecutionContext
-
-case class ClusterIamSetupException(googleProject: GoogleProject)
-    extends LeoException(s"Error occurred setting up IAM roles in project ${googleProject.value}")
+import scala.concurrent.{ExecutionContext, Future}
 
 class ClusterHelper(
   dbRef: DbReference,
   dataprocConfig: DataprocConfig,
+  googleGroupsConfig: GoogleGroupsConfig,
   gdDAO: GoogleDataprocDAO,
   googleComputeDAO: GoogleComputeDAO,
+  googleDirectoryDAO: GoogleDirectoryDAO,
   googleIamDAO: GoogleIamDAO
 )(implicit val executionContext: ExecutionContext, val system: ActorSystem, val contextShift: ContextShift[IO])
     extends LazyLogging
     with Retry {
 
   def createClusterIamRoles(googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo): IO[Unit] =
-    updateClusterIamRoles(googleProject, serviceAccountInfo, true)
+    updateClusterIamRoles(googleProject, serviceAccountInfo, createCluster = true)
 
   def removeClusterIamRoles(googleProject: GoogleProject, serviceAccountInfo: ServiceAccountInfo): IO[Unit] =
-    updateClusterIamRoles(googleProject, serviceAccountInfo, false)
+    updateClusterIamRoles(googleProject, serviceAccountInfo, createCluster = false)
+
+  def generateServiceAccountKey(googleProject: GoogleProject,
+                                serviceAccountEmailOpt: Option[WorkbenchEmail]): IO[Option[ServiceAccountKey]] =
+    // TODO: implement google2 version of GoogleIamDAO
+    serviceAccountEmailOpt.traverse { email =>
+      IO.fromFuture(IO(googleIamDAO.createServiceAccountKey(googleProject, email)))
+    }
+
+  def removeServiceAccountKey(googleProject: GoogleProject,
+                              serviceAccountEmailOpt: Option[WorkbenchEmail],
+                              serviceAccountKeyIdOpt: Option[ServiceAccountKeyId]): IO[Unit] =
+    // TODO: implement google2 version of GoogleIamDAO
+    (serviceAccountEmailOpt, serviceAccountKeyIdOpt).mapN {
+      case (email, keyId) =>
+        IO.fromFuture(IO(googleIamDAO.removeServiceAccountKey(googleProject, email, keyId)))
+    } getOrElse IO.unit
+
+  def setupDataprocImageGoogleGroup(): IO[Boolean] =
+    createDataprocImageUserGoogleGroupIfItDoesntExist()
+      .flatMap { _ =>
+        addIamRoleToDataprocImageGroup(dataprocConfig.customDataprocImage)
+      }
 
+  /**
+   * Add the user's service account to the Google group.
+   * This group has compute.imageUser role on the custom Dataproc image project,
+   * which allows the user's cluster to pull the image.
+   */
+  def updateDataprocImageGroupMembership(googleProject: GoogleProject, createCluster: Boolean): IO[Unit] =
+    dataprocConfig.customDataprocImage.flatMap(parseImageProject) match {
+      case None => IO.unit
```

Comment:
1. you can use `.traverse`
2. I'd get rid of 
```
              val projectNumberOptIO = IO.fromFuture(IO(googleComputeDAO.getProjectNumber(googleProject)))
              val clusterIamSetupExceptionIO = IO.raiseError[Long](ClusterIamSetupException(imageProject))
```

Something like
```
    dataprocConfig.customDataprocImage.flatMap(parseImageProject).traverse {
      imageProject =>
        for {
          count <- IO.fromFuture(IO(dbRef.inTransaction {
            _.clusterQuery.countActiveByProject(googleProject)
          }))
          _ <- // Note: Don't remove the account if there are existing active clusters in the same project,
            // because it could potentially break other clusters. We only check this for the 'remove' case.
            if (count > 0 && !createCluster) {
              IO.unit
            } else {
              for {
                projectNumberOptIO <- IO.fromFuture(IO(googleComputeDAO.getProjectNumber(googleProject)))
                projectNumber <- IO.fromEither(projectNumberOptIO.toRight(ClusterIamSetupException(imageProject)))
                // Note that the Dataproc service account is used to retrieve the image, and not the user's
                // pet service account. There is one Dataproc service account per Google project. For more details:
                // https://cloud.google.com/dataproc/docs/concepts/iam/iam#service_accounts
                dataprocServiceAccountEmail = WorkbenchEmail(
                  s"service-${projectNumber}@dataproc-accounts.iam.gserviceaccount.com"
                )
                _ <- updateGroupMembership(googleGroupsConfig.dataprocImageProjectGroupEmail,
                  dataprocServiceAccountEmail,
                  createCluster)
              } yield ()
            }
        } yield ()
    }.void
```

Reply:
+1